### PR TITLE
P4 spec issue 1261 add bounded loops

### PR DESCRIPTION
**Pull request by jafingerhut**

### CODE REVIEW

**Comment by jafingerhut** [comment]
This is NOT a complete implementation of for loops in p4c.  It probably isn't even a complete or correct change to the syntax as of the first 2 commits.  It is a work in progress that I have used to verify that with these proposed grammar changes, bison does not give an error message when it processes the grammar.

**Comment by asl** [comment]
@jafingerhut Some time ago we implemented bounded loops downstream. They employ `for (variable in sequence)` syntax as a way to enforce that the loop is indeed bounded (it seems quite a non-trivial task in general for C-style `for (start; end; increment)` loops).

Let me know if you'd like me to factor our this change and submit just to have some end-to-end implementation.

**Comment by fruffy** [comment]
Bounded loops are a top-level agenda item of the P4 language working group for this year:

https://docs.google.com/document/d/1XSgdXeG1UuF1FM_XAqxDrHeN4dHZWBnJPKVS6SnGNwM/edit#heading=h.q7bzcwglv61g

Upstreaming this is a great idea and very useful. How the syntax will end up looking like is a different problem though.

**Comment by asl** [comment]
Sure, I can see what could be done.

**Comment by jafingerhut** [comment]
@asl Thanks for the offer.  Fabian gave the high order answer.  There should be a public meeting focused on the topic of loops in P4 coming up soon-ish (within a month or so), and I can be sure to try to include you in the choice of meeting time so it fits in your schedule.

File: frontends/parsers/p4/p4parser.ypp [hunk 1]
```diff
@@ -1283,6 +1290,49 @@ statementOrDeclaration
...
+forUpdateStatementsNonEmpty
+    : assignmentOrMethodCallStatementWithoutSemicolon
+    | assignmentOrMethodCallStatementWithoutSemicolon ","
+      forUpdateStatementsNonEmpty
+    ;
+
+forCollectionExpr
+    : expression
+    | expression ".." expression
+    | typeRef
```
Comment (ChrisDodd) [inline frontends/parsers/p4/p4parser.ypp:1333]: What is a `typeRef` supposed to indicate here?  Creating a temp of the type and then iterating over it?

**Review by ChrisDodd — COMMENTED** [review]
I've started looking at loops in the compiler, so I'm looking at what to add in the IR for these, and what is needed in other parts of the compiler.

**Comment by ChrisDodd** [comment]
I added an extra commit with the IR changes for this.

**Comment by asl** [comment]
> I added an extra commit with the IR changes for this.

I've almost finished porting our changes for the loops. I will post a PR shortly

**Comment by asl** [comment]
I ported our downstream code at https://github.com/p4lang/p4c/pull/4558

Currently use-def changes are missed, I am working on porting them asap.

Let me know if there are any questions / comments / suggestions

**Comment by ChrisDodd** [comment]
I've pulled together both changes in #4562

**Comment by asl** [comment]
I was not part of the original discussion, so do not know about all the details here. But my understanding is that if the intention is to support bounded loops, then I think it should iterate over collection (or as in my original PR, over explicit range like `a ... b`). Generic C-style loops `for (start; cond; update)` are much harder to prove that they are bounded.

**Comment by jafingerhut** [comment]
Closing this preliminary PR, as https://github.com/p4lang/p4c/pull/4562 and other changes by Chris Dodd have been merged in.

**Pull request closed without merging**